pylinalg/func/vector.py

```diff
@@ -116,3 +116,114 @@ def vector_apply_matrix(vectors, matrix, /, *, w=1, out=None, dtype=None):
     if dtype is not None:
         out = out.astype(dtype, copy=False)
     return out
+
+
+def project_camera(vector, projection_matrix) -> np.ndarray:
+    """Project a 3D vector into 2D space.
+
+    Parameters
+    ----------
+    vector : ArrayLike
+        The vector to be projected.
+    projection_matrix: ArrayLike
+        The camera's intrinsic matrix.
+
+    Returns
+    -------
+    projected_vector : ndarray
+        The projection of the vector.
+
+    """
+
+    raise NotImplementedError()
+
+
+def unproject_camera(vector, projection_matrix, /, *, depth=0) -> np.ndarray:
+    """Un-project a vector from 2D space to 3D space.
+
+    Parameters
+    ----------
+    vector : ArrayLike
+        The vector to be un-projected.
+    projection_matrix: ArrayLike
+        The camera's intrinsic matrix.
+    depth : float
+        The distance of the unprojected vector from the camera.
```

Comment:
I'd say 0..1 would map from near to far?

Reply:
Usually there are three matrices worth mentioning when rendering a 3D model to the screen.

World (or Model) matrix: transforms a model's vertices to world space.

View matrix: transforms world space coordinates to camera space. That means they're positioned "in front of" the camera, and "upright", with the camera at the origin.

Projection matrix: transforms view space coordinates to clip space/NDC (normalized device) coordinates. Typically there is some notion of a camera frustum which defines the transform. Coordinates that lie within the camera frustum end up on the [-1, 1] range on all three axes.

After this, the GPU takes over. Coordinates within the [-1, 1] clipping range are rendered onto the screen.

I probably got a bunch of details wrong but at least it gives you some handles to google for more info. ;)

Reply:
> Usually there are three matrices worth mentioning when rendering a 3D model to the screen.

Ah okay. That indeed clarifies things; thanks! In robotics, we call the view matrix the _extrinsic matrix_ and the projection matrix the _intrinsic matrix_. Once we know the intrinsic matrix (through camera calibration), we can use reference points to estimate the extrinsic matrix and, once that is known, we can estimate World/Model matrices.

> I'd say 0..1 would map from the near to the far plane? Do you have a different idea?

I wouldn't limit the `depth` to 0-1. It would reduce the generality of the function and it's easy enough to pass `depth=z*(far - near)` to `unproject` and get the desired result.

Reply:
>  the [-1, 1] range on all three axes.

Worth mentioning that in OpenGL depth lies in  -1..1, but on wgpu it's 0..1.

> View matrix: transforms world space coordinates to camera space. That means they're positioned "in front of" the camera, and "upright", with the camera at the origin.

In PyGfx the view matrix is literally the inverse of the camera's object transform.